### ISSUE TITLE
Making 'static BrokenLinkCallback

### ISSUE THREAD

**Issue by ollpu** [raphlinus/pulldown-cmark#697]
I'm having trouble constructing a `Parser` with a broken link callback such that `'callback: 'static`. My use case is to use a parser in a function with the following signature:

```rust
fn parse(input: &str) -> impl Iterator<Item = Event<'_>>
```

However, if I want to add a broken link callback, I am unable to satisfy that signature.

For example:

```rust
fn callback<'input>(_link: BrokenLink<'input>) -> Option<(CowStr<'input>, CowStr<'input>)> {
    todo!()
}

fn parse(input: &str) -> impl Iterator<Item = Event<'_>> {
    Parser::new_with_broken_link_callback(&input, Options::empty(), Some(&mut callback))
}
```

gives

```
error[E0716]: temporary value dropped while borrowed
  --> src/lib.rs:32:79
   |
31 | fn parse(input: &str) -> impl Iterator<Item = Event<'_>> {
   |                 - let's call the lifetime of this reference `'1`
32 |     Parser::new_with_broken_link_callback(&input, Options::empty(), Some(&mut callback))
   |     --------------------------------------------------------------------------^^^^^^^^--
   |     |                                                                         |
   |     |                                                                         creates a temporary value which is freed while still in use
   |     argument requires that borrow lasts for `'1`
33 | }
   | - temporary value is freed at the end of this statement

For more information about this error, try `rustc --explain E0716`.
```

I tried some tricks, but I couldn't come up with a way to construct a `&'static mut dyn FnMut()`. Going through `const _: fn() = callback` gives a similar error because of the temporary created at the use site. A `static` item cannot be directly borrowed mutably, so it can't be put in a `&mut` without some temporary wrapper.

Using `static mut` works but requires unsafe. If there's a way to do this without unsafe I'd be interested to know!

I believe a more flexible API would be to accept `Option<Box<dyn FnMut(..) + 'callback>>` instead of `Option<&'callback mut dyn FnMut(..)>`. A value of the latter can be converted to the former by simply boxing the mutable reference, as they transitively implement `FnMut`. Performance might be slightly worse, but that seems negligible.

The BrokenLinkCallback API was reworked previously in #509. Changing to boxes would be yet another breaking change unfortunately.

**Comment by ollpu** [comment]
Reading #509 more carefully, the trait + generics approach suggested there would also address this issue.

**Comment by Martin1887** [comment]
So can this be closed?

Thanks.

**Comment by ollpu** [comment]
That API change was discussed in #509 but not implemented. So instead of having a `dyn FnMut` closure, `Parser` would be generic over a callback trait.

I suppose this issue can be closed if the intent of #508 is to move to generics, which would solve both issues. The approach hasn't been discussed there though, only in the later closed issue.

**Comment by Martin1887** [comment]
OK, this is an enhancement then and it will be fixed in some months because there are other things with higher priority. Feel free to create a pull request to speed up the process.

Thanks.

**Comment by Martin1887** [comment]
This has been fixed by #701, right?

**Comment by ollpu** [comment]
Ah, yes, didn't remember to tag it.
